### PR TITLE
Added thread existence check at thread_wakeup

### PR DESCRIPTION
**Pull request by IldarValiev**
Current version doesn't check if pid is valid and can damage innocent data.

### CODE REVIEW

**Comment by Kijewski** [comment]
The else-case for `if (other_thread->status == STATUS_SLEEPING) {` should return `0` not `STATUS_NOT_FOUND` if this change is accepted.

**Comment by kaspar030** [comment]
Hm, so in order to be able to print the DEBUG, we duplicate the code for irq restoring.
Can't we just, if DEBUG is enabled, print the right thing depending on whether other_thread==NULL?

**Comment by IldarValiev** [comment]
@Kijewski, according to description, everything is correct

Returns
1 on success
STATUS_NOT_FOUND if pid is unknown or not sleeping

**Comment by IldarValiev** [comment]
@kaspar030, this will return one condition in if-case and add one "if" in else-case

**Comment by kaspar030** [comment]
@IldarValiev But the "if" in else case would be within ```#ifdef ENABLE_DEBUG```. 

Also, when ENABLE_DEBUG is defined, the compiler will pretty sure test for "other_thread==NULL" once and then jump directly into the "if" block within the "else" block. And it would save the duplicate "restoreIRQ(); return STATUS_NOT_FOUND" lines.

If I read your patch correctly, when "ENABLE_DEBUG" (which conditionally enables DEBUG statements) is not set, STATUS_NOT_FOUND is returned - just as before.

**Comment by kaspar030** [comment]
@IldarValiev Better. Could you confirm that this results in the same code when ENABLE_DEBUG is 0?

File: core/thread.c [hunk 1]
```diff
@@ -71,8 +71,12 @@ int thread_wakeup(kernel_pid_t pid)
 
     unsigned old_state = disableIRQ();
 
-    tcb_t *other_thread = (tcb_t *) sched_threads[pid];
-    if (other_thread && other_thread->status == STATUS_SLEEPING) {
+    tcb_t *other_thread = thread_get(pid);
+    
+    if (!other_thread){
```
Comment (cgundogan) [inline core/thread.c]: minor: could you add a whitespace after the closing parentheses please?

**Comment by IldarValiev** [comment]
@kaspar030 Yes, I can, but not right know. Currently I don't even have any compiler on my current machine. Give me a couple of days to install all necessary software, please.

**Comment by Kijewski** [comment]
> @Kijewski, according to description, everything is correct

Sorry, I did not explain myself well enough. I meant that if the change is accepted, then the API should be changed as well to distinguish between an invalid target and a target that is not sleeping.

**Comment by IldarValiev** [comment]
@Kijewski, I don't understand why this function returns hardcoded values in some cases. What is "1"? Does it have "STATUS_OK" meaning? If yes, then why is it hardcoded? Why should it return "0" in case of the thread is not sleeping? Is it "STATUS_ERROR"? 

I think that return values in this file (and may be in the whole core) are need to be reviewed.

**Comment by Kijewski** [comment]
Essentially I am proposing: http://thedailywtf.com/articles/What_Is_Truth_0x3f_ :)

**Comment by IldarValiev** [comment]
@kaspar030, this won't produce the same (binary) code, because now here is included one function call and "if-elseif-else" introduced instead of "if-else".

**Comment by kaspar030** [comment]
@IldarValiev I like the other optimization you've added.

Would you mind splitting the commit in to? And please prepend the commit message with "core: thread:"

**Comment by IldarValiev** [comment]
@kaspar030, done.

**Comment by OlegHahm** [comment]
@IldarValiev, oops, this got lost, would you mind to rebase?

**Comment by IldarValiev** [comment]
Sorry for the delay and for deleting of all history in this PR - I tried to become friends with Git for Windows. It was harder than I thought. :)

Hope now everything is ok.

**Comment by OlegHahm** [comment]
> This branch has conflicts that must be resolved

Hm, either you missed something or it needs yet another rebase.

**Comment by IldarValiev** [comment]
Rebased. The core/thread.c file in master branch was changed yesterday.

**Comment by OlegHahm** [comment]
Travis complains:
```
core/thread.c:149: tab in indent.
+       if (stacksize < 0) {
core/thread.c:75: error (noQuoteCharPair): No pair for character ("). Can't process file. File is either invalid or unicode, which is currently not supported.
```

**Comment by IldarValiev** [comment]
Fixed.

File: core/thread.c [hunk 2]
```diff
@@ -69,8 +69,12 @@ int thread_wakeup(kernel_pid_t pid)
 
     unsigned old_state = disableIRQ();
 
-    thread_t *other_thread = (thread_t *) sched_threads[pid];
-    if (other_thread && other_thread->status == STATUS_SLEEPING) {
+    thread_t *other_thread = thread_get(pid);
+
+    if (!other_thread) {
+        DEBUG("thread_wakeup: Thread does not exist!".\n");
```
Comment (kaspar030) [inline core/thread.c]: typo

**Comment by IldarValiev** [comment]
Fixed typo.

File: core/thread.c [hunk 3]
```diff
@@ -69,8 +69,12 @@ int thread_wakeup(kernel_pid_t pid)
 
     unsigned old_state = disableIRQ();
 
-    thread_t *other_thread = (thread_t *) sched_threads[pid];
-    if (other_thread && other_thread->status == STATUS_SLEEPING) {
+    thread_t *other_thread = thread_get(pid);
```
Comment (OlegHahm) [inline core/thread.c]:
```
 /home/travis/build/RIOT-OS/RIOT/core/thread.c: In function 'thread_wakeup':
/home/travis/build/RIOT-OS/RIOT/core/thread.c:72:30: error: initialization discards 'volatile' qualifier from pointer target type [-Werror]
     thread_t *other_thread = thread_get(pid);
```

**Comment by OlegHahm** [comment]
@kaspar030, did you ACK?

**Comment by OlegHahm** [comment]
@kaspar030, @gebart, @Kijewski, anyone else?

**Comment by authmillenon** [comment]
ping?

**Comment by kaspar030** [comment]
The stacksize check is unrelated, but .....

ACK.

**Comment by authmillenon** [comment]
`mutex_unlock_and_sleep` (which incidentally is using `thread_wakeup()`). So I give the second ACK.

**Pull request merged**